File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamnotifications/callback/OfflineMessageNotificationCallback.java
package in.dragonbra.javasteam.steam.handlers.steamnotifications.callback;

import in.dragonbra.javasteam.protobufs.steamclient.SteammessagesClientserver2.CMsgClientOfflineMessageNotification;
import in.dragonbra.javasteam.steam.handlers.steamnotifications.SteamNotifications;
import in.dragonbra.javasteam.steam.steamclient.callbackmgr.CallbackMsg;
import in.dragonbra.javasteam.types.SteamID;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Fired in response to calling {@link SteamNotifications#requestOfflineMessageCount()}.
 */
public class OfflineMessageNotificationCallback extends CallbackMsg {

    private int messageCount;

    private List<SteamID> friendsWithOfflineMessages;

    public OfflineMessageNotificationCallback(CMsgClientOfflineMessageNotification.Builder msg) {
        messageCount = msg.getOfflineMessages();

        friendsWithOfflineMessages = new ArrayList<>();
        for (Integer id : msg.getFriendsWithOfflineMessagesList()) {
            friendsWithOfflineMessages.add(new SteamID(id));
        }
        friendsWithOfflineMessages = Collections.unmodifiableList(friendsWithOfflineMessages);
    }

    /**
     * @return the number of new messages
     */
    public int getMessageCount() {
        return messageCount;
    }

    /**
     * @return the ids of friends the new messages belong to
     */
    public List<SteamID> getFriendsWithOfflineMessages() {
        return friendsWithOfflineMessages;
    }
}

File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamnotifications/callback/OfflineMessageNotificationCallback.kt
package `in`.dragonbra.javasteam.steam.handlers.steamnotifications.callback

import `in`.dragonbra.javasteam.base.ClientMsgProtobuf
import `in`.dragonbra.javasteam.base.IPacketMsg
import `in`.dragonbra.javasteam.protobufs.steamclient.SteammessagesClientserver2.CMsgClientOfflineMessageNotification
import `in`.dragonbra.javasteam.steam.handlers.steamnotifications.SteamNotifications
import `in`.dragonbra.javasteam.steam.steamclient.callbackmgr.CallbackMsg
import `in`.dragonbra.javasteam.types.SteamID

/**
 * Fired in response to calling [SteamNotifications.requestOfflineMessageCount].
 */
class OfflineMessageNotificationCallback(packetMsg: IPacketMsg) : CallbackMsg() {

    /**
     * Get the number of new messages
     */
    val messageCount: Int

    /**
     * Gets the ids of friends the new messages belong to
     */
    val friendsWithOfflineMessages: List<SteamID>

    init {
        val resp = ClientMsgProtobuf<CMsgClientOfflineMessageNotification.Builder>(
            CMsgClientOfflineMessageNotification::class.java,
            packetMsg
        )
        val msg = resp.body

        messageCount = msg.offlineMessages
        friendsWithOfflineMessages = msg.friendsWithOfflineMessagesList.map { SteamID(it.toLong()) }
    }
}
